server/server.go

```diff
@@ -2187,7 +2189,7 @@ func findBestQueueSub(sl []*subState) (rsub *subState) {
 
 		// Favor non stalled subscribers and clients that do not have
 		// failed heartbeats
-		if (!sStalled || rStalled) && (!sHasFailedHB || rHasFailedHB) && (sOut < rOut) {
+		if (!sStalled && !sHasFailedHB && (rStalled || rHasFailedHB)) || (sOut < rOut) {
```

Comment:
But for this PR and to maintain current behavior, we may want to pick the first at the end of the loop if `rsub` has not been set in the loop.